Macos progress bar

adomasven opened the pull request:
Closes #1923 

The issue with displaying the progress bar on macOS is that when you close it (to then open a citation dialog), the main Zotero window is raised (appears in the foreground with the citation dialog focused on top of it). This is because when you close a window of an application on macOS it automatically focuses the last focused window of the app, which in our case after closing the progress bar is the Zotero window.

In the past we haven't been showing the citation dialog on macOS due to this, as it's quite jarring, especially if the Zotero window raises under the quick format dialog

This PR makes it so that Refresh, Unlink Citations and Add/Edit Bibliography show a progress bar. After Refresh and Unlink Citations is finished, Document.activate is called, which means that Word receives focus and the Zotero window is not raised. For Add/Edit Bibliography, if a bibliography already exists, then the Edit Bibliography dialog is displayed, with the Zotero library window raised, which is not perfect, but fine, since Edit Bibliography dialog is large. Since Add/Edit Bibliography refreshes the document and may take a while, this is still better than no progress bar.

For Add/Edit Citation, Add Note and Doc Prefs we do not display an initial progress bar, because these operations are generally faster and only require fetching doc prefs to open their respective dialogs. A progress bar will be displayed after confirming these dialog. This brings the Zotero window into the foreground, but if you are changing citation styles or such, it's much more important to know when the progress has finished than to not get any progress dialog at all.

Note that after all these operations finish, we call Document.active, which focuses the relevant doc editor, so in the end the fact that the Zotero window was raised doesn't cause huge issues.

I believe, that a variety of "intermittent" errors that we see from macOS are actually errors from users who are having their documents updated with no indication (progress bar) from Zotero, and removing or editing citations while Zotero is working on their document. This change should make it much easier for us to help or troubleshoot the errors for those users, as well as generally improve the usability of the integration experience with larger docs where operations may take a long time but display no indication that Zotero is working.

abaevbog reviewed (commented):
Everything is working as expected for me! These are just a couple of small interface-related suggestions:

abaevbog commented on chrome/content/zotero/integration/quickFormat.js:
```diff
@@ -1524,9 +1524,12 @@ var Zotero_QuickFormat = new function () {
-		if(percent === null) {
+		meter.style.display = "";
+		await Zotero.Promise.delay(10);
+		_resizeWindow(meter.parentElement.parentElement.getBoundingClientRect().height);
```
For me, the deck (`meter.parentElement.parentElement`) retains the height of the `editor`. If there are many lines of bubbles, `meter.parentElement.parentElement.getBoundingClientRect().height` is the same as the current height, so this line does not change the window height for me.

<img width="796" alt="Screenshot 2025-01-29 at 9 43 16 AM" src="https://github.com/user-attachments/assets/f55d75e7-ded5-4e0d-8e0d-fc18bdc186d1" />

Do we want to resize the window to only have enough space for the progress bar? We could run `editor.replaceChildren()` in `accept` to remove all bubbles after the citation object is updated to let `editor` (and the parent deck) go back to its original single-line height?

adomasven replied:
I'm generally not happy with the need to rely on the XUL deck here, since it seems to replace the items in some async way. I guess it's slower when there's more elements. But it's easy to switch to a more reliable approach.

abaevbog commented on chrome/skin/default/zotero/integration.css:
```diff
@@ -216,8 +216,12 @@ span.zotero-bubble-input {
+	border-radius: 10px;
+	margin: 0;
+	flex: 1;
+	height: 10px;
```
Should we also add `align-self: center;` so that the progress bar is in the center of the dialog, as opposed to being pushed into the upper-half?

<img width="799" alt="Screenshot 2025-01-29 at 10 17 16 AM" src="https://github.com/user-attachments/assets/126c556a-b00c-4511-aab4-cc7beaf0107d" />

adomasven replied:
Not relevant since the dialog just has to (and will now) resize to only contain the progress bar.

abaevbog reviewed (commented):
Everything is working as expected for me! These are just a couple of small interface-related suggestions:

adomasven commented:
@abaevbog addressed your comments. The dialog will now correctly resize when finalizing and displaying the progress bar so that only it is visible which was the intention in the first place but the deck element was acting weird.

abaevbog commented:
Great, yeah it all works well for me now!
